site/en/docs/extensions/reference/declarativeNetRequest/index.md

```diff
@@ -158,102 +164,139 @@ examples of URL filters:
+### Rule limits {: #limits }
 
-When an extension is queried for how to handle a request, the highest priority matching rule is returned. If more than one matching rule has the highest priority, the tie is broken
-based on the action type, in the following order of decreasing precedence:
+There is a performance overhead to loading and evaluating rules in the browser,
+and so a number of limits apply when using the API. Limits depend on the type of
+rule you're using.
+
+#### Static rules {: #static-rules }
 
-- `allow`
-- `allowAllRequests`
-- `block`
-- `upgradeScheme`
-- `redirect`
+Static rules are those specified in rule files (which are specified in the manifest file).
 
-If the request was not blocked or redirected, the matching `modifyHeaders` rules are evaluated with
-the most recently installed extensions getting priority. Within each extension, all `modifyHeaders`
-rules with a priority lower than matching `allow` or `allowAllRequests` rules are ignored.
+An extension can specify up to **50** static [rulesets](#type-Ruleset) as part of the `"rule_resources"` manifest key, but only **10** of these rulesets can be enabled at a time. The latter is called the [`MAX_NUMBER_OF_ENABLED_STATIC_RULESETS`](#property-MAX_NUMBER_OF_ENABLED_STATIC_RULESETS). Collectively, those rulesets are guaranteed at least **30,000** rules. This is called the [`GUARANTEED_MINIMUM_STATIC_RULES`](#property-GUARANTEED_MINIMUM_STATIC_RULES). The number of rules available after that depends on how many rules are enabled by all the extensions installed on a user's browser. You can find this number at runtime by calling [`getAvailableStaticRuleCount()`](#method-getAvailableStaticRuleCount). You can see [an example of this](#update-static-rulesets) under [code examples](#code-examples).
 
-If multiple `modifyHeaders` rules specify the same header, the resulting modification for the header
-is determined based on the priority of each rule and the operations specified.
+#update-static-rulesets
 
-- If a rule has appended to a header, then lower priority rules can only append to that header.
-  `set` and `remove` operations are not permitted.
-- If a rule has set a header, then lower priority rules cannot further modify the header, except for
-  `append` rules from the same extension.
-- If a rule has removed a header, then lower priority rules cannot further modify the header.
+#### Dynamic and session rules {: #dynamic-session-rules }
+
+The limits applied to dynamic and session rules are simpler than static rules. The total number of both cannot exceed **5000**. This is called the [`MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES`][#property-MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES].
+
+#### Rules that use regex {: #regex-rules }
+
+All types of rules can use regular expressions; however the total number of regex rules, of each type cannot exceed **1000**. This is called the [MAX_NUMBER_OF_REGEX_RULES](#property-MAX_NUMBER_OF_REGEX_RULES).
```

Comment:
There's also a memory limit on RegEx rules (I believe 2KB) that we should check and document.

Reply:
Confirmed that the memory is limited to 2KB here: https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:extensions/browser/api/declarative_net_request/utils.cc;l=317;drc=0340dd86c6c1f01ccd1a0b983633cfef5fcf6697

You get an error that looks like this (which is somewhat helpful, although it doesn't tell you what "complex" means behind the scenes 😔):

<img width="525" alt="Screenshot 2023-08-01 at 11 03 21" src="https://github.com/GoogleChrome/developer.chrome.com/assets/6097064/d4d7ec01-2f2b-4a10-89ae-8294f3ddf0c0">

Reply:
I'm open to wording suggestions.

Reply:
> Additionally, each rule must be less than 2KB once compiled. This correlates with the general complexity of the rule specified. If you try to load a rule which exceeds this limit, you will see a warning and the rule will be ignored.

First draft to work from :)

Reply:
Added the paragraph, plus an example of the warning (in case someone google's the text of the warning).

Reply:
Love it!